### DISCUSSION TITLE
Move IJobFactory to IJobRunShellFactory?

### DISCUSSION THREAD

**Discussion by drieseng** [quartznet/quartznet#1772] [Ideas]
We only need **IJobFactory** when we produce an **IJob** instance in `JobRunShell.Initialize(QuartzScheduler sched, CancellationToken cancellationToken = default)]`.

Should we consider removing the **JobFactory** property from **QuartzScheduler** and instead "inject" it into the **IJobRunShellFactory**. The concrete **IJobRunShellFactory** instance would in turn pass the **IJobFactory** onto the **JobRunShell** by adding an extra argument to the `JobRunShell(IScheduler scheduler, TriggerFiredBundle bundle)` ctor or via an extra argument to the `JobRunShell.Initialize(QuartzScheduler sched, CancellationToken cancellationToken = default)` method.

We would of course have to update **StdSchedulerFactory** and **DirectSchedulerFactory** accordingly.

**Note:**
If you do not reject the idea immediately, I'd be happy to implement it in a PR knowing that it may still be rejected after all.

**Comment by lahma** [comment]
Generally I would push the idea of components getting their dependencies via constructor inject. One of v4 goals is to have MS DI handling dependencies when possible. So no interfaces or other setter-like ways to give dependencies, just a "promise" about getting them injected via constructor when injected interface is known as available and implementation present.

Not sure if I answered the question though 😉

**Reply by lahma** [reply]
And as always, all help is much appreciated. We probably need to have good game plan before implementation so that it aligns with the goals.
